src/antidote_crdt_set_rw.erl

```diff
@@ -73,82 +73,123 @@
     | {remove_all, [member()]}
     | {reset, {}}.
 
--type token() :: term().
--type effect() ::
-      {add, [{member(), RemTokens::[token()]}], ResetTokens::[token()]}
-    | {remove, [{member(), RemTokens::[token()]}], NewRemToken::token(), ResetTokens::[token()]}
-    | {reset, [{member(), RemTokens::[token()]}], OldResetTokens::[token()], NewResetToken::token()}.
+%% element, CurrentTokens, AddTokens, RemoveTokens
+-type downstream_op() :: [{member(), tokens(), tokens(), tokens()}].
 
 -type member() :: term().
+-type token() :: term().
+-type tokens() :: [token()].
 
 -spec new() -> set().
 new() ->
-  {[], orddict:new()}.
+  orddict:new().
 
 -spec value(set()) -> [member()].
-value({_, Dict}) ->
-    [Val || {Val, []} <- orddict:to_list(Dict)].
-
-
--spec downstream(set_op(), set()) -> {ok, effect()}.
-downstream({add, Elem}, State) ->
-    downstream({add_all, [Elem]}, State);
-downstream({add_all, Elems}, {ResetTokens, Dict}) ->
-    {ok, {add, elems_with_removed_tombstones(lists:usort(Elems), Dict, ResetTokens), ResetTokens}};
-downstream({remove, Elem}, State) ->
-    downstream({remove_all, [Elem]}, State);
-downstream({remove_all, Elems}, {ResetTokens, Dict}) ->
-    {ok, {remove, elems_with_removed_tombstones(lists:usort(Elems), Dict, ResetTokens), unique(), ResetTokens}};
-downstream({reset, {}}, {ResetTokens, Dict}) ->
-  RemovedTombstones = elems_with_removed_tombstones(lists:usort(orddict:fetch_keys(Dict)), Dict, ResetTokens),
-  {ok, {reset, RemovedTombstones, ResetTokens, unique()}}.
+value(RWSet) ->
+    orddict:fold(
+        fun(Elem, {AddTokens, RemoveTokens}, Acc) ->
+            case AddTokens =/= [] andalso RemoveTokens == [] of
+                true ->
+                    Acc ++ [Elem];
+                false ->
+                    Acc
+            end
+        end,
+        [],
+        RWSet
+    ).
+
+-spec downstream(set_op(), set()) -> {ok, downstream_op()}.
+downstream({add, Elem}, RWSet) ->
+  downstream({add_all, [Elem]}, RWSet);
+downstream({add_all, Elems}, RWSet) ->
+  CreateDownstream = fun(Elem, CurrentTokens) ->
+      Token = unique(),
+      {Elem, CurrentTokens, [Token], []}
+  end,
+  DownstreamOps = create_downstreams(CreateDownstream, lists:usort(Elems), RWSet, []),
+  {ok, lists:reverse(DownstreamOps)};
+downstream({remove, Elem}, RWSet) ->
+    downstream({remove_all, [Elem]}, RWSet);
+downstream({remove_all, Elems}, RWSet) ->
+  CreateDownstream = fun(Elem, CurrentTokens) ->
+      Token = unique(),
+      {Elem, CurrentTokens, [], [Token]}
+  end,
+  DownstreamOps = create_downstreams(CreateDownstream, lists:usort(Elems), RWSet, []),
+  {ok, lists:reverse(DownstreamOps)};
+downstream({reset, {}}, RWSet) ->
+  CreateDownstream = fun(Elem, CurrentTokens) ->
+      {Elem, CurrentTokens, [], []}
+  end,
+  DownstreamOps = create_downstreams(CreateDownstream, lists:usort(orddict:fetch_keys(RWSet)), RWSet, []),
+  {ok, lists:reverse(DownstreamOps)}.
```

Comment:
yeah, as per our discussion, we can remove these by fixing create_downstreams() in both set_rw and orset. Fow now, I kept them the same. Later, we can fix them both files and move the commonalities to a different module.